lti_consumer/plugin/views.py

```diff
@@ -501,3 +556,430 @@ def list(self, *args, **kwargs):
+        TieredCache.set_all_tiers(end_assessment_return_key, end_assessment_return_value)
+
+    return JsonResponse(data={})
+
+
+# We do not want Django's CSRF protection enabled for POSTs made by external services to this endpoint. This is because
+# Django uses the double-submit cookie method of CSRF protection, but the Proctoring Specification lends itself better
+# to the synchronizer token method of CSRF protection. Django's method requires an anti-CSRF token to be included in
+# both a cookie and a hidden from value in the request to CSRF procted endpoints. In the Proctoring Specification, there
+# are a number of issues supporting the double-submit cookie method.
+#
+# 1. Django requires that a cookie containing the anti-CSRF token is sent with the request from the Proctoring Tool to
+#    the Assessment Platform . When the user's browser makes a request to the launch_gate_endpoint view, an anti-CSRF
+#    token is set in the cookie. The default SameSite attribute for cookies is "Lax" (stored in the Django setting
+#    CSRF_COOKIE_SAMESITE), meaning that when the Proctoring Tool redirects the user's browser back to the Assessment
+#    Platform, the browser will not include the previously set cookie in its request to the Assessment Platform.
+#    CSRF_COOKIE_SAMESITE can be set to "None", but this means that all third-party cookies will be included by the
+#    browser, which may compromise CSRF protection for other endpoints. Note that settings CSRF_COOKIE_SAMESITE to
+#    "None" requires that CSRF_COOKIE_SECURE is set to True.
+#
+# 2. Django validates a request by comparing the above anti-CSRF token in the cookie to the anti-CSRF token in the POST
+#    request parameters. Django expects the anti-CSRF token to be in the POST request parameters with the key name
+#    "csrfmiddlewaretoken". However, the Proctoring Specification requires that the anti-CSRF token be included in the
+#    JWT token with the name "session_data". The Proctoring Tool will not direct the browser to send this anti-CSRF
+#    token back with the key name "csrfmiddlewaretoken", nor will it include it as a form parameter, as it's not part of
+#    the Proctoring Services Specification.
+@csrf_exempt
+# Authorization Servers MUST support the use of the HTTP GET and POST methods defined in RFC 2616 [RFC2616]
+# at the Authorization Endpoint.
+# See 3.1.2.1.Authentication Request of the OIDC specification.
+# https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+@require_http_methods(["GET", "POST"])
+def launch_gate_endpoint_proctoring(request, suffix=None):  # pylint: disable=unused-argument
+    """
+    This is the Assessment Platform's OIDC login authentication/authorization endpoint.
+
+    This uses the "client_id" query parameter or form data to identify the LtiConfiguration and its consumer to generate
+    the LTI 1.3 Launch Form.
+    """
+    preflight_response = request.GET.dict() if request.method == 'GET' else request.POST.dict()
```

Comment:
this is super weird, there isn't a request method that just does this? request.THING_YOU_HAVE.dict()?

Reply:
I couldn't find one, no ☹️ .